cpu/cc2538/spi-arch.h

```diff
@@ -34,63 +34,173 @@
+ *
+ * This SPI driver depends on the following defines:
+ *
+ * For the SSI0 module:
+ *
+ * - SPI0_CKL_PORT
+ * - SPI0_CLK_PIN
+ * - SPI0_TX_PORT
+ * - SPI0_TX_PIN
+ * - SPI0_RX_PORT
+ * - SPI0_RX_PIN
+ *
+ * For the SSI1 module:
+ *
+ * - SPI1_CKL_PORT
+ * - SPI1_CLK_PIN
+ * - SPI1_TX_PORT
+ * - SPI1_TX_PIN
+ * - SPI1_RX_PORT
+ * - SPI1_RX_PIN
  */
 #ifndef SPI_ARCH_H_
 #define SPI_ARCH_H_
 
 #include "dev/ssi.h"
+/*---------------------------------------------------------------------------*/
+/* The default SPI instance to use. You can configure either instance 0 or 1
+   and proceed to use the "old" function calls / macros */
+#ifdef SPI_CONF_DEFAULT_INSTANCE
+#define SPI_DEFAULT_INSTANCE      SPI_CONF_DEFAULT_INSTANCE
+#else
+#define SPI_DEFAULT_INSTANCE      0
+#endif
+/*---------------------------------------------------------------------------*/
+/* We define a new set of "SPIX" macros which will then be mapped to the
+   corresponding macros defined later on */
+#define SPIX_WAITFORTxREADY(x)      SPI##x##_WAITFORTxREADY()
+#define SPIX_WAITFOREOTx(x)       SPI##x##_WAIT_FOREOTx()
+#define SPIX_WAITFOREORx(x)       SPI##x##_WAITFOREORx()
+#define SPIX_FLUSH(x)         SPI##x##_FLUSH()
+/*---------------------------------------------------------------------------*/
+/* Deprecated macros provided for compatibility reasons */
+#if (SPI_DEFAULT_INSTANCE == 0)
+#define SPI_WAITFORTxREADY()      SPIX_WAITFORTxREADY(0)
+#define SPI_TXBUF           SPI0_TXBUF
+#define SPI_RXBUF           SPI0_RXBUF
+#define SPI_WAITFOREOTx()       SPIX_WAITFOREOTx(0)
+#define SPI_WAITFOREORx()       SPIX_WAITFOREORx(0)
+#ifdef SPI_FLUSH
+#error "You must include spi-arch.h before spi.h for the CC2538."
+#else
+#define SPI_FLUSH()           SPIX_FLUSH(0)
+#endif
+#elif (SPI_DEFAULT_INSTANCE == 1)
+#define SPI_WAITFORTxREADY()      SPIX_WAITFORTxREADY(1)
+#define SPI_TXBUF           SPI1_TXBUF
+#define SPI_RXBUF           SPI1_RXBUF
+#define SPI_WAITFOREOTx()       SPIX_WAITFOREOTx(1)
+#define SPI_WAITFOREORx()       SPIX_WAITFOREORx(1)
+#ifdef SPI_FLUSH
+#error "You must include spi-arch.h before spi.h for the CC2538."
+#else
+#define SPI_FLUSH()           SPIX_FLUSH(1)
+#endif
+#else
+#error "Invalid SPI instance. Valid values are 0 or 1"
+#endif
 
-#define SPI_WAITFORTxREADY() do { \
-  while(!(REG(SSI0_BASE + SSI_SR) & SSI_SR_TNF)); \
+#define SPI_CS_CLR(port, pin)     SPIX_CS_CLR(port, pin)
+#define SPI_CS_SET(port, pin)     SPIX_CS_SET(port, pin)
+/*---------------------------------------------------------------------------*/
+/* New API macros */
+#define SPI0_WAITFORTxREADY() do { \
+    while(!(REG(SSI0_BASE + SSI_SR) & SSI_SR_TNF)) ; \
 } while(0)
-
-#define SPI_TXBUF REG(SSI0_BASE + SSI_DR)
-
-#define SPI_RXBUF REG(SSI0_BASE + SSI_DR)
-
-#define SPI_WAITFOREOTx() do { \
-  while(REG(SSI0_BASE + SSI_SR) & SSI_SR_BSY); \
+#define SPI0_TXBUF REG(SSI0_BASE + SSI_DR)
+#define SPI0_RXBUF REG(SSI0_BASE + SSI_DR)
```

Comment:
It's the same register, and these definitions are internal to the driver (they're exposed, but not supposed to be used elsewhere), so merge them like this:
```
#define SPIX_BUF(spi) REG(TPASTE3(SSI, spi, _BASE) + SSI_DR)
```